<a href="https://colab.research.google.com/github/almc6742/cu-git-workshop/blob/master/Exercise_3_LASSO_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import pandas
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LassoLarsCV
from sklearn.datasets import make_regression
import matplotlib.pyplot as plt
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [0]:
sales_data = 'drive/My Drive/Colab Notebooks/finalmaster-ratios.csv'
df = pd.read_csv(sales_data)

List of predictors:

In [0]:
allvariablenames = list(df.columns.values)

Remove first 8 variables:

In [0]:
listofallpredictors = allvariablenames[8:]

Get columns ready in pandas:

In [0]:
predictors = df[listofallpredictors]

target = df['# Purchases']

Split the data (by rows):

In [0]:
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123)

Use LassoLarsCV to build a LASSO model:

In [0]:
X = predictors
Y = target

model = LassoLarsCV(cv=10, precompute=False).fit(X, Y)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 8 iterations, i.e. alpha=8.526e-01, with an active set of 8 regressors, and the smallest cholesky pivot element being 8.816e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 12 iterations, i.e. alpha=4.263e-01, with an active set of 12 regressors, and the smallest cholesky pivot element being 8.816e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 12 iterations, i.e. alpha=4.263e-01, with an active set of 12 regressors, and the smallest cholesky pivot element

In [0]:
predictors_model=pd.DataFrame(listofallpredictors)
predictors_model.columns = ['label']
predictors_model['coeff'] = model.coef_

for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)

['B01001014' 0.5330720656233966]
['B01001036' 1.6591227573088478]
['B01001037' 2.4482060101637564]
['B01001038' 2.21381020716577]
['B02001005' 0.9859155344708008]
['B02001007' 0.006591090821753925]
['B13014016' 0.0061795097515955225]
['B13014026' 0.574990933842321]
['B13014027' 1.2849909864550721]
['B13016008' 0.31692716289106393]
['B15002027' 3.9418447021477347]
['B19001017' 1.8860806124452758]


Question #1:
The first line of code creates a Pandas dataframe from the predictor variables in our dataset. The second line of code initalizes an indexable column (label) for the variable names. The third line of code pulls the coefficients from our LASSO model as a second column aligned with their respective predictor variable column name. The "for-loop" iterates over the rows in our newly created dataframe and keeps all significant rows, or rows in which the coefficients are greater than 0. The result is an array containing which variables most predicted sales along with their associated coefficients.

Question #2: The greatest predictor of sales is the census variable B15002027, respresenting females in the population aged 25 and older whom completed the 12th grade but did not receive a diploma. Based on the provided predictor variables, we sell the most Bobo Bars among this 25+ female population with no high school diploma but completed education through 12th grade. The next two strongest predictors include females aged 30-34 followed by females 35-39.

Question #3:
The two census variables that most steeply predicted sales include B15002027 and B01001036.

In [0]:
from sklearn.metrics import mean_squared_error

train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

training data MSE
21060.761034858362


In [0]:
test_error = mean_squared_error(tar_test, model.predict(pred_test))
print ('testing data MSE')
print(test_error)

testing data MSE
36561.48369809522


Question #4:
Using the MSE tool from sklearn, we find that the test set has a mean-squared error of 36561.484 while the training set has a mean-squared error of 21060.761. There is greater bias and variance in our test set than our training set; however, n is large enough such that the difference is negligible. In practical terms, determining the MSE helped us to assess the closeness of our predictor variable data to the target variable when bias and sampling error are taken into account. 

In [0]:
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.2733096196457573


In [0]:
rsquared_test=model.score(pred_test,tar_test)
print ('testing data R-square')
print(rsquared_test)

testing data R-square
0.27480133108928484


Question #5:
The R-squared of the training set is nearly equivalent to that of the test set. The census data predicts a significant amount of overall sales data. The predictor variables in this dataset can be used to predict sales in the future.

In [0]:
print("y interecept:")
print(model.intercept_)

y interecept:
-3.9606645084610435


Question #6:
Our baselines sales number is essentially 4 purchases (above 3.96). This means that 4 sales must be made before we can say with certainty that our model will help predict sales based on provided data.